pkg/scheduler/framework/v1alpha1/context.go

```diff
@@ -32,22 +33,41 @@ type ContextData interface{}
 // ContextKey is the type of keys stored in PluginContext.
 type ContextKey string
 
+var _ context.Context = &PluginContext{}
+
 // PluginContext provides a mechanism for plugins to store and retrieve arbitrary data.
 // ContextData stored by one plugin can be read, altered, or deleted by another plugin.
 // PluginContext does not provide any data protection, as all plugins are assumed to be
 // trusted.
 type PluginContext struct {
+	context.Context
```

Comment:
The intention is that `PluginContext` implements `context.Context`?
As such, it looks like this is what built-in contexts do https://golang.org/src/context/context.go#L492

However, I would say that we should get rid of types `ContextKey` and `ContextData`, as well as method `Read`. This would makes this class closer to a builtin context.

Reply:
I like the idea, but it looks like we need the access to the functions to set values? Do we actually need them to be dynamic though? Maybe we can just set them at creation time, just like built-in contexts.

Reply:
> Do we actually need them to be dynamic though? Maybe we can just set them at creation time, just like built-in contexts.

`pluginContext` is designed for passing parameters through different plugins

Reply:
alright, then I don't think we should return `context.Context`. That will just make things harder down the line.

Reply:
We could return `pluginContext` which conformance `context.Context` interface and use `context.Context` as the first argument of `schedule`

Reply:
Yes, but when plugins need to write values to it, then they have to cast back to the internal type (or an interface to it). In that case, I prefer that we keep the explicit type.

Reply:
That would prevent devs from accidentally providing another context implementation or wrapping the pluginContext.

Reply:
@Huang-Wei what do you think of the approach above?

Reply:
The root `context.Context` is instantiated from the up-most caller, i.e. `cmd/*`, so as long as we honor the context derivation hierachy I'm fine.

@draveness Let me confirm, upon the instantiation of `pluginContext`, we embed the derived `context.Context` and keep using that down the way, right? For other occurances, .e.g, in the upper callers, they're not aware of `pluginContext`, so still use `context.Context`, right?

BTW: in https://github.com/kubernetes/kubernetes/pull/82072#discussion_r321530394, my intention was to avoid "embedded context", per the Golang guide.

Reply:
I took another look at the Golang guide, if we follow it strictly we should neither create a custom context (without embedded context) nor use embedded context (in the initial approach). The only way seems valid could be passed context and plugin context to all the function like this:

```go
func schedule(ctx context.Context, pluginContext framework.PluginContext, pod *v1.Pod)
```

I'm not sure I like this approach, passing two contexts to a function seems weird, maybe we could rename pluginContext to pluginMetadata.

Reply:
+1 to having 2 parameters and rename the type to `PluginMetadata` or simply `Metadata`

Reply:
+1 to 2 parameters unless it involves way too many efforts. (in that case we can fall back to the embedded solution)

Renaming `pluginContext` is necessary if we pass along context.Context, @ahg-g any naming suggestion here?

Reply:
done